packages/playwright/src/plugins/webServerPlugin.ts

```diff
@@ -102,14 +118,35 @@ export class WebServerPlugin implements TestRunnerPlugin {
       cwd: this._options.cwd,
       stdio: 'stdin',
       shell: true,
-      // Reject to indicate that we cannot close the web server gracefully
-      // and should fallback to non-graceful shutdown.
-      attemptToGracefullyClose: () => Promise.reject(),
+      attemptToGracefullyClose: async () => {
+        if (process.platform === 'win32')
+          throw new Error('Graceful shutdown is not supported on Windows');
+        if (!signal)
+          throw new Error('skip graceful shutdown');
+
+        const success = launchedProcess.kill(signal);
+        if (!success)
+          throw new Error(`signal didn't succeed, fall back to non-graceful shutdown`);
+
+        return new Promise<void>((resolve, reject) => {
+          const timer = timeout !== 0
+            ? setTimeout(() => {
+              // @ts-expect-error. SIGINT didn't kill the process, but `processLauncher` will only attempt killing it if this is false
+              launchedProcess.killed = false;
```

Comment:
Yeah agree. It might be fine today, but who knows how `node:child_process` changes. I've refactored it in https://github.com/microsoft/playwright/pull/33379/commits/220ed29dbeb38fe6900a1a6d5509eb47f072a61f, how do you like that?